analyzers/tests/SonarAnalyzer.UnitTest/TestCases/TimestampsShouldNotBeUsedAsPrimaryKeys.cs

```diff
@@ -0,0 +1,139 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+using DateTimeAlias = System.DateTime;
+using KeyAttributeAlias = System.ComponentModel.DataAnnotations.KeyAttribute;
+
+class TemporalTypes
+{
+    class Entity
+    {
+        public DateTime EntityId { get; set; }          // Noncompliant {{Timestamps should not be used as primary keys}}
+        //     ^^^^^^^^
+    }
+
+    class DateTimeKey
+    {
+        public DateTime Id { get; set; }                // Noncompliant
+    }
+
+    class DateTimeOffsetKey
+    {
+        public DateTimeOffset Id { get; set; }          // Noncompliant
+    }
+
+    class TimeSpanKey
+    {
+        public TimeSpan Id { get; set; }                // Noncompliant
+    }
+
+    class DateTimeNoKey
+    {
+        public DateTime Identifier { get; set; }        // Compliant - only Id and [class-name]Id is recognized as key by Entity Framework
+    }
+
+    class DifferentCasing
+    {
+        public DateTime ID { get; set; }                // Noncompliant
+    }
+
+    class TemporalTypeWithFullName
+    {
+        public System.DateTime Id { get; set; }         // Noncompliant
+    }
+
+    class AliasedTemporalType
+    {
+        public DateTimeAlias Id { get; set; }           // FN - the chance of using a type aliased temporal type as a database key is slim, so we don't cover it to improve performance
+    }
+}
+
+class NonTemporalTypes
+{
+    class IntKey
+    {
+        public int Id { get; set; }                     // Compliant - not a temporal type
+    }
+
+    class GuidKey
+    {
+        public Guid Id { get; set; }
+    }
+
+    class StringKey
+    {
+        public string Id { get; set; }
+    }
+}
+
+class Attributes
+{
+    class ClassWithSingleKeyAttribute
```

Comment:
You can add also a UT with `[PrimaryKey]` attribute ([it's a thing](https://learn.microsoft.com/en-us/ef/core/modeling/keys?tabs=data-annotations#configuring-a-primary-key) starting from EF Core 7)

Reply:
While it's possible to create single-property keys with it, `[PrimaryKey]` was mainly introduced to create composite keys.
I don't think the rule should raise a warning when a `DateTime` is part of a composite key:

```cs
[PrimaryKey(nameof(DateProperty), nameof(IntProperty))]
class PrimaryKeyWithSingleProperty
{
    public DateTime DateProperty { get; set; }  // Compliant - the rule will not raise warnings when a temporal type is part of a composite key
    public int IntProperty { get; set; }
}
```

We don't really know the context, and how the developer guarantees uniqueness when the key is composed of multiple values. So I opt not to raise an issue here.
As for using `PrimaryKey` with a single property: it's possible, but it's not what it was introduced for. I've seen relatively few examples for this on SourceGraph. For now I've documented it as a FN. I'll add it as optional task, If there's enough time for it, then I'll do it after the VB implementation is done, in a separate PR.